File: 02-Arrays-Part-II/java/05-FindRepeatingAndMissingNumbers.java
class Solution {
    public int[] findMissingRepeatingNumbers(int[] nums) {
        int n = nums.length;

        int xr = 0;
        for (int i = 0; i < n; i++) {
            xr ^= nums[i];
            xr ^= (i + 1);
        }

        int number = xr & ~(xr - 1);

        int zero = 0, one = 0;

        for (int i = 0; i < n; i++) {
            if ((nums[i] & number) != 0) {
                one ^= nums[i];
            } else {
                zero ^= nums[i];
            }
        }

        for (int i = 1; i <= n; i++) {
            if ((i & number) != 0) {
                one ^= i;
            } else {
                zero ^= i;
            }
        }

        int cnt = 0;
        for (int val : nums) {
            if (val == zero) cnt++;
        }

        if (cnt == 2) {
            return new int[]{zero, one};
        }

        return new int[]{one, zero};
    }
}

File: 02-Arrays-Part-II/python/05_find_repeating_and_missing_numbers.py
class Solution:
    def findMissingRepeatingNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums)

        xr = 0
        for i in range(n):
            xr ^= nums[i]
            xr ^= (i + 1)

        number = xr & ~(xr - 1)

        zero, one = 0, 0

        for i in range(n):
            if nums[i] & number:
                one ^= nums[i]
            else:
                zero ^= nums[i]

        for i in range(1, n + 1):
            if i & number:
                one ^= i
            else:
                zero ^= i

        cnt = sum(1 for val in nums if val == zero)

        if cnt == 2:
            return [zero, one]

        return [one, zero]
